### PR TITLE
login: T4751: 2FA OTP key generator in VyOS CLI

### PR DESCRIPTION
**Pull request by goodNETnick**
## Change Summary

Added ability to generate OTP keys for login from VyOS CLI.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T4751

## Component(s) name

system, login, ssh

## Proposed changes

Added operation mode command:
generate system login username otptester otp-key hotp-time <parameters>

## How to test

```
vyos@vyos:~$ generate system login username otptester otp-key hotp-time rate-limit 2 rate-time 20 window-size 5
# You can share it with the user, he just needs to scan the QR in his OTP app
# username:  otptester
# OTP KEY:  J5A64ERPMGJOZXY6FMHHLKXKANNI6TCY
# OTP URL:  otpauth://totp/otptester@vyos?secret=J5A64ERPMGJOZXY6FMHHLKXKANNI6TCY&digits=6&period=30
█████████████████████████████████████████████
█████████████████████████████████████████████
████ ▄▄▄▄▄ █▀█ █▄   ▀▄▀▄█▀▄  ▀█▀ █ ▄▄▄▄▄ ████
████ █   █ █▀▀▀█ ▄▀ █▄▀ ▀▄ ▄ ▀  ▄█ █   █ ████
████ █▄▄▄█ █▀ █▀▀██▄▄ █ █ ██ ▀▄▀ █ █▄▄▄█ ████
████▄▄▄▄▄▄▄█▄▀ ▀▄█ █ ▀ █ █ █ █▄█▄█▄▄▄▄▄▄▄████
████ ▄   █▄ ▄ ▀▄▀▀▀▀▄▀▄▀▄▄▄▀▀▄▄▄  █ █▄█ █████
████▄▄ ██▀▄▄▄▀▀█▀ ▄ ▄▄▄ ▄▀ ▀ █ ▄ ▄ ██▄█  ████
█████▄  ██▄▄▀█▄█▄█▄ ▀█▄▀▄ ▀█▀▄ █▄▄▄ ▄   ▄████
████▀▀▄   ▄█▀▄▀ ▄█▀█▀▄▄▄▀█▄ ██▄▄▄  ▀█ █  ████
████ ▄▀▄█▀▄▄█▀▀▄▀▀▀▀█ ▄▀▄▀ ▄█ ▀▄  ▄ ▄▀ █▄████
████▄ ██ ▀▄▀▀ ▄█▀ ▄ ██ ▀█▄█ ▄█ ▄ ▀▄   ▄▄ ████
████▄█▀▀▄ ▄▄ █▄█▄█▄ █▄▄▀▄▄▀▀▄▄██▀ ▄▀▄▄ ▀▄████
████▀▄▀ ▄ ▄▀█ ▄ ▄█▀ █  ▀▄▄  ▄█▀ ▄▄   ▀▄▄ ████
████  ▀███▄ █▄█▄▀▀▀▀▄ ▄█▄▄▀ ▀███ ▄▄█▄▄  ▄████
████ ███▀ ▄▄▀▀██▀ ▄▀▄█▄▄▄ ██▄▄▀▄▀  ███▄ ▄████
████▄████▄▄▄▀▄ █▄█▄▀▄▄▄▄██▀ ▄▀ ▄ ▄▄▄ █▄▄█████
████ ▄▄▄▄▄ █▄▄▄ ▄█▀█▀▀▀▀█▀█▀ █▄█ █▄█ ▄█  ████
████ █   █ █ ██▄▀▀▀▀▄▄▄▀ ▄▄▄  ▀ ▄    ▄ ▄▄████
████ █▄▄▄█ █ ▀▀█▀ ▄▄█ █▄▄██▀▀█▀ █▄▀▄██▄█ ████
████▄▄▄▄▄▄▄█▄█▄█▄█▄▄▄▄▄█▄▄▄█▄██████▄██▄▄▄████
█████████████████████████████████████████████
█████████████████████████████████████████████
# To add this OTP key to configuration, run the following commands:
set system login user otptester authentication otp key 'J5A64ERPMGJOZXY6FMHHLKXKANNI6TCY'
set system login user otptester authentication otp rate-limit '2'
set system login user otptester authentication otp rate-time '20'
set system login user otptester authentication otp window-size '5'

```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly

### CODE REVIEW

**Comment by sever-sever** [comment]
Could you set commit message without white spaces at the start
I mean not use “system login: Txxx”
Use “login: Txxx” or just “Txxx: foo descr”
As commit message doesn’t pass

**Comment by goodNETnick** [comment]
> Could you set commit message without white spaces at the start I mean not use “system login: Txxx” Use “login: Txxx” or just “Txxx: foo descr” As commit message doesn’t pass

Thanks for the clarification
Fixed

File: op-mode-definitions/generate-system-login-user.xml.in [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                  <node name="otp-key">
+                    <properties>
+                      <help>Generate OpenConnect OTP token</help>
+                    </properties>
+                    <children>
+                      <node name="hotp-time">
+                        <properties>
+                          <help>HOTP time-based token</help>
+                        </properties>
+                        <command>sudo ${vyos_op_scripts_dir}/generate_system_login_user.py --username "$5" --rate_limit 3 --rate_time 30 --window_size 3</command>
```
Comment (c-po) [inline op-mode-definitions/generate-system-login-user.xml.in]: Why have default values here and also in `generate_system_login_user.py`?

File: op-mode-definitions/generate-system-login-user.xml.in [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                        <children>
+                          <tagNode name="rate-limit">
+                            <properties>
+                              <help>Duration of single time interval</help>
+                            </properties>
+                            <command>sudo ${vyos_op_scripts_dir}/generate_system_login_user.py --username "$5" --rate_limit "$9" --rate_time 30 --window_size 3</command>
```
Comment (c-po) [inline op-mode-definitions/generate-system-login-user.xml.in]: Why have default values here and also in `generate_system_login_user.py`?

File: op-mode-definitions/generate-system-login-user.xml.in [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                            <children>
+                              <tagNode name="rate-time">
+                                <properties>
+                                  <help>The number of digits in the one-time password</help>
+                                </properties>
+                                <command>sudo ${vyos_op_scripts_dir}/generate_system_login_user.py --username "$5" --rate_limit "$9" --rate_time "${11}" --window_size 3</command>
```
Comment (c-po) [inline op-mode-definitions/generate-system-login-user.xml.in]: Why have default values here and also in `generate_system_login_user.py`?

File: op-mode-definitions/generate-system-login-user.xml.in [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                                  </tagNode>
+                                </children>
+                              </tagNode>
+                            </children>
+                          </tagNode>
+                          <tagNode name="window-size">
+                            <properties>
+                              <help>The number of digits in the one-time password</help>
+                            </properties>
+                            <command>sudo ${vyos_op_scripts_dir}/generate_system_login_user.py --username "$5" --rate_limit "3" --rate_time "30" --window_size "${9}"</command>
```
Comment (c-po) [inline op-mode-definitions/generate-system-login-user.xml.in]: Why have default values here and also in `generate_system_login_user.py`?

**Review by c-po — CHANGES_REQUESTED** [review]

File: op-mode-definitions/generate-system-login-user.xml.in (hunk 1 above)
Comment (goodNETnick) [reply op-mode-definitions/generate-system-login-user.xml.in]: This script can be used for a variety of purposes
Flexibility is not a bad thing ;)
But this is not too important. Removed defaults from XML
Fixed

File: op-mode-definitions/generate-system-login-user.xml.in (hunk 2 above)
Comment (goodNETnick) [reply op-mode-definitions/generate-system-login-user.xml.in]: Fixed

File: op-mode-definitions/generate-system-login-user.xml.in (hunk 1 above)
Comment (goodNETnick) [reply op-mode-definitions/generate-system-login-user.xml.in]: Fixed

File: op-mode-definitions/generate-system-login-user.xml.in (hunk 3 above)
Comment (goodNETnick) [reply op-mode-definitions/generate-system-login-user.xml.in]: Fixed

File: op-mode-definitions/generate-system-login-user.xml.in (hunk 4 above)
Comment (goodNETnick) [reply op-mode-definitions/generate-system-login-user.xml.in]: Fixed

**Pull request merged**